### PR TITLE
Linux supervisor

### PR DESCRIPTION
**Pull request by mdomasevicius**
Enable linux installation & daemon/client transport

### CODE REVIEW

**Review by vkuznecovas — APPROVED** [review]

File: supervisor/daemon/transport/transport_linux.go [hunk 1]
```diff
@@ -17,10 +17,56 @@
...
+		return fmt.Errorf("error listening: %w", err)
+	}
+	numUid, err := strconv.Atoi(options.Uid)
+	if err != nil {
+		return fmt.Errorf("failed to parse uid %s: %w", options.Uid, err)
+	}
+	if err := os.Chown(sock, numUid, -1); err != nil {
+		return fmt.Errorf("failed to chown supervisor socket to uid %s: %w", options.Uid, err)
+	}
+	if err := os.Chmod(sock, 0700); err != nil {
```
Comment (mdomasevicius) [inline supervisor/daemon/transport/transport_linux.go:48]: I am not sure what permissions should be applied here, but desktop app does not start unless it's `766` I believe

Comment (tomasmik) [reply supervisor/daemon/transport/transport_linux.go:48]: As I understand desktop app is running in a different user space, so ye `0700` would not work.

File: supervisor/install/install_linux.go [hunk 2]
```diff
@@ -18,15 +18,108 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	t, err := template.New("unit-descriptor").Parse(description)
+	if err != nil {
+		return nil, err
+	}
+	buffer := new(bytes.Buffer)
+	err = t.Execute(buffer, map[string]string{
```
Comment (soffokl) [inline supervisor/install/install_linux.go:85]: Where this template used? It's just created and never reused or I'm missing something?

File: supervisor/install/install_linux.go [hunk 3]
```diff
@@ -18,15 +18,108 @@
...
+
+func clean(d daemon.Daemon) error {
+	execAndLog(d.Stop)
+	return execAndLog(d.Remove)
+}
+
+func execAndLog(action func() (string, error)) error {
+	output, err := action()
+	if err != nil {
+		log.Info().Msgf("%s\t%s", output, err)
```
Comment (soffokl) [inline supervisor/install/install_linux.go:110]: Should it be an `Error` or `Warn` if `err != nil`?
```suggestion
		log.Error().Msgf("%s\t%s", output, err)
```

File: supervisor/install/install_linux.go [hunk 4]
```diff
@@ -18,15 +18,108 @@
...
+		return err
+	}
+	log.Info().Msg(output)
+	return nil
 }
 
-// Uninstall installs service for linux. Not implemented yet.
+// Uninstall - remove supervisor daemon
```
Comment (soffokl) [inline supervisor/install/install_linux.go:117]:
```suggestion
// Uninstall removes supervisor daemon.
```

File: supervisor/daemon/wireguard/wginterface/interface_linux.go [hunk 5]
```diff
@@ -18,20 +18,59 @@
...
+	"github.com/rs/zerolog/log"
+	"golang.zx2c4.com/wireguard/device"
+	"golang.zx2c4.com/wireguard/ipc"
 
 	"golang.zx2c4.com/wireguard/tun"
 )
 
 func createTunnel(requestedInterfaceName string) (tunnel tun.Device, interfaceName string, err error) {
-	return nil, requestedInterfaceName, errors.New("not implemented")
+	log.Info().Msg("createTunnel start")
```
[the remark this answers is not in the document]
Comment (mdomasevicius) [reply supervisor/daemon/wireguard/wginterface/interface_linux.go:35]: forgot to remove, these were for debugging

File: supervisor/install/install_linux.go [hunk 6]
```diff
@@ -18,15 +18,108 @@
...
+	log.Info().Msg("Cleaning up previous installation")
+	clean(dmn)
+
+	log.Info().Msg("Installing daemon")
+	err = execAndLog(func() (string, error) { return dmn.Install() })
+	if err != nil {
+		return err
+	}
+
+	err = execAndLog(dmn.Start)
```
[the remark this answers is not in the document]
Comment (mdomasevicius) [reply supervisor/install/install_linux.go:67]: This library prints return unix like output

```
output, err := action()
	if err != nil {
		log.Error().Msgf("%s\t%s", output, err)
		return err
	}
```

e.g.: `Starting Daemon        [ OK ]` 

So I figured that extra "Starting daemon" is not necessary

File: supervisor/install/install_linux.go [hunk 7]
```diff
@@ -18,15 +18,108 @@
...
+	dmn, err := mystSupervisorDaemon(options)
+	if err != nil {
+		return err
+	}
+
+	log.Info().Msg("Cleaning up previous installation")
+	clean(dmn)
+
+	log.Info().Msg("Installing daemon")
+	err = execAndLog(func() (string, error) { return dmn.Install() })
```
[the remark this answers is not in the document]
Comment (mdomasevicius) [reply supervisor/install/install_linux.go]:
```console
supervisor/install/install_linux.go:62:22: cannot use dmn.Install (type func(...string) (string, error)) as type func() (string, error) in argument to execAndLog
```

it has different signature

image

File: supervisor/install/install_linux.go (hunk 2 above)
Comment (mdomasevicius) [reply supervisor/install/install_linux.go:82]: yes, a bug - lost some changes with stashing

File: supervisor/client/transport_linux.go [hunk 8]
```diff
@@ -18,10 +18,17 @@
 package client
 
 import (
-	"errors"
+	"fmt"
 	"io"
+	"net"
 )
 
+const sock = "/var/run/myst.sock"
```
[the remark this answers is not in the document]
Comment (mdomasevicius) [reply supervisor/client/transport_linux.go:26]: Should this be aligned with `transport/transport_linux.go` `const sock = "/var/run/myst.sock"` ?

[the remark this answers is not in the document]
Comment (soffokl) [reply supervisor/client/transport_linux.go:26]: Yep, it's kind of will work even without it, but `/var/run/` is legacy and just a symlink for compatibility for old systems.

**Review by soffokl — APPROVED** [review]

**Pull request merged**